### PR TITLE
fieldpath: add `DeleteField` function to delete elements from Paved

### PR DESCRIPTION
**Pull request by muvaf**
### Description of your changes

Introduces `DeleteField` to delete a field from the given object.

Fixes https://github.com/crossplane/crossplane-runtime/issues/343

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Unit tests with 100% coverage.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by turkenh — APPROVED** [review]

**Comment by muvaf** [comment]
I've noticed that two cases weren't handled and added them:
* Top-level field could not be removed since iteration assumed there is at least two segments, i.e. accessing `i+1`. Now we handle this as special case.
* In cases where the given path does not exist even before the last node, i.e. `metadata.some.another.other` where there is no `metadata.some`, it returned error. Now it's no-op since the target path is already gone.

Added unit tests for both but couldn't reduce cyclomatic complexity without breaking `p.delete` into pieces that don't make much sense.

@turkenh could you take another look?

File: pkg/fieldpath/paved.go [hunk 1]
```diff
@@ -471,3 +471,113 @@ func (p *Paved) SetBool(path string, value bool) error {
 func (p *Paved) SetNumber(path string, value float64) error {
 	return p.SetValue(path, value)
 }
+
+// DeleteField deletes the field from the object.
+// If the path points to an entry in an array, the element
+// on that index is removed and the next ones are pulled
+// back. If it is a field on a map, the field is
+// removed from the map.
+func (p *Paved) DeleteField(path string) error {
```
Comment (muvaf) [inline pkg/fieldpath/paved.go:480]: BTW, I could be convinced to rename this to `DeleteKey`.

**Review by turkenh — APPROVED** [review]

File: pkg/fieldpath/paved.go (hunk 1 above)
Comment (turkenh) [reply pkg/fieldpath/paved.go:480]: `Key` wouldn't be a good fit if array. What about `DeleteItem` ?

Comment (negz) [reply pkg/fieldpath/paved.go:480]: Would simply `Delete` work? As in `paved.Delete` to delete whatever is at the supplied field path?

Comment (muvaf) [reply pkg/fieldpath/paved.go:480]: @turkenh I feel like `Field` is more familiar than `Item` in this context for the users of fieldpath.

@negz I think it'd read nice if we had `Get`/`Set`/`Delete` but given that there is `GetValue`/`SetValue`, `paved.Delete` feels a bit off without a noun that says what you're deleting.

Thanks for the suggestions folks!

**Pull request merged**